frame/cere-staking/src/lib.rs

```diff
@@ -0,0 +1,599 @@
+	/// set, it might be active or not.
+	#[pallet::storage]
+	#[pallet::getter(fn current_era)]
+	pub type CurrentEra<T> = StorageValue<_, EraIndex>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An account has bonded this amount. \[stash, amount\]
+		///
+		/// NOTE: This event is only emitted when funds are bonded via a dispatchable. Notably,
+		/// it will not be emitted for staking rewards when they are added to stake.
+		Bonded(T::AccountId, BalanceOf<T>),
+		/// An account has unbonded this amount. \[stash, amount\]
+		Unbonded(T::AccountId, BalanceOf<T>),
+		/// An account has called `withdraw_unbonded` and removed unbonding chunks worth `Balance`
+		/// from the unlocking queue. \[stash, amount\]
+		Withdrawn(T::AccountId, BalanceOf<T>),
+		/// An account has stopped participating as either a storage network or CDN participant.
+		/// \[stash\]
+		Chilled(T::AccountId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Not a controller account.
+		NotController,
+		/// Not a stash account.
+		NotStash,
+		/// Stash is already bonded.
+		AlreadyBonded,
+		/// Controller is already paired.
+		AlreadyPaired,
+		/// Cannot have a storage network or CDN participant, with value less than the minimum
+		/// defined by governance (see `MinStorageBond` and `MinEdgeBond`). If unbonding is the
+		/// intention, `chill` first to remove one's role as storage/edge.
+		InsufficientBond,
+		/// Can not schedule more unlock chunks.
+		NoMoreChunks,
+		/// Internal state has become somehow corrupted and the operation cannot continue.
+		BadState,
+		// An account already declared a desire to participate in the network with a certain role
+		// and to take another role it should call `chill` first.
+		AlreadyInRole,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Take the origin account as a stash and lock up `value` of its balance. `controller` will
+		/// be the account that controls it.
+		///
+		/// `value` must be more than the `minimum_balance` specified by `T::Currency`.
+		///
+		/// The dispatch origin for this call must be _Signed_ by the stash account.
+		///
+		/// Emits `Bonded`.
+		#[pallet::weight(100_000)]
+		pub fn bond(
+			origin: OriginFor<T>,
+			controller: <T::Lookup as StaticLookup>::Source,
+			#[pallet::compact] value: BalanceOf<T>,
+		) -> DispatchResult {
+			let stash = ensure_signed(origin)?;
+
+			if <Bonded<T>>::contains_key(&stash) {
+				Err(Error::<T>::AlreadyBonded)?
+			}
+
+			let controller = T::Lookup::lookup(controller)?;
+
+			if <Ledger<T>>::contains_key(&controller) {
+				Err(Error::<T>::AlreadyPaired)?
+			}
+
+			// Reject a bond which is considered to be _dust_.
+			if value < T::Currency::minimum_balance() {
+				Err(Error::<T>::InsufficientBond)?
+			}
+
+			frame_system::Pallet::<T>::inc_consumers(&stash).map_err(|_| Error::<T>::BadState)?;
```

Comment:
What is `inc_consumers` responsible for?

Reply:
It increments a reference counter related to the given account to ensure another pallet does not delete it until our pallet uses it.

> The consumers reference counter indicating the number of other modules that currently depend on this account's existence.

More: https://docs.substrate.io/reference/account-data-structures#account-reference-counters